src/cudamatrix/cu-sparse-matrix.cc

```diff
@@ -70,7 +71,8 @@ template <typename Real>
 Real CuSparseMatrix<Real>::FrobeniusNorm() const {
 #if HAVE_CUDA == 1
   if (CuDevice::Instantiate().Enabled()) {
-    CuVector<Real> element_vec(*this);
+    CuVector<Real> element_vec(this->NumElements(), kUndefined);
+    this->CopyElementsToVec(&element_vec);
     return element_vec.Norm(2);
```

Comment:
OK I merged master and unit tests passed.

Guoguo

On Tue, Jan 5, 2016 at 4:08 PM, Daniel Povey <notifications@github.com>
wrote:

> In src/cudamatrix/cu-sparse-matrix.cc
> <https://github.com/kaldi-asr/kaldi/pull/434#discussion_r48913949>:
>
> > @@ -70,7 +71,8 @@ template <typename Real>
> >  Real CuSparseMatrix<Real>::FrobeniusNorm() const {
> >  #if HAVE_CUDA == 1
> >    if (CuDevice::Instantiate().Enabled()) {
> > -    CuVector<Real> element_vec(*this);
> > +    CuVector<Real> element_vec(this->NumElements(), kUndefined);
> > +    this->CopyElementsToVec(&element_vec);
> >      return element_vec.Norm(2);
>
> oh sorry, I forgot the definition.
> … <#152144356c736f1f_>
> On Tue, Jan 5, 2016 at 4:06 PM, Guoguo Chen <notifications@github.com>
> wrote: In src/cudamatrix/cu-sparse-matrix.cc <#434 (comment)
> <https://github.com/kaldi-asr/kaldi/pull/434#discussion_r48913804>>: > @@
> -70,7 +71,8 @@ template <typename Real> > Real
> CuSparseMatrix<Real>::FrobeniusNorm() const { > #if HAVE_CUDA == 1 > if
> (CuDevice::Instantiate().Enabled()) { > - CuVector<Real>
> element_vec(*this); > + CuVector<Real> element_vec(this->NumElements(),
> kUndefined); > + this->CopyElementsToVec(&element_vec); > return
> element_vec.Norm(2); We are computing sqrt(sum_{i=1}^N (x_i^2)) as the
> Frobenius norm right? Why would the number of elements matter? Perhaps I
> misunderstood CuVector.Norm()? Guoguo … <#1521441390d26195_> On Tue, Jan 5,
> 2016 at 3:31 PM, Daniel Povey <notifications@github.com> wrote: In
> src/cudamatrix/cu-sparse-matrix.cc <#434
> <https://github.com/kaldi-asr/kaldi/pull/434> (comment) <#434 (comment)
> <https://github.com/kaldi-asr/kaldi/pull/434#discussion_r48911401>>>: >
> @@ -70,7 +71,8 @@ template <typename Real> > Real
> CuSparseMatrix<Real>::FrobeniusNorm() const { > #if HAVE_CUDA == 1 > if
> (CuDevice::Instantiate().Enabled()) { > - CuVector<Real>
> element_vec(*this); > + CuVector<Real> element_vec(this->NumElements(),
> kUndefined); > + this->CopyElementsToVec(&element_vec); > return
> element_vec.Norm(2); This code is not right, you need account for the fact
> that the matrix's size is different from the number of nonzero elements.
> Otherwise it's not really the frobenius norm. — Reply to this email
> directly or view it on GitHub <
> https://github.com/kaldi-asr/kaldi/pull/434/files#r48911401>. — Reply to
> this email directly or view it on GitHub <
> https://github.com/kaldi-asr/kaldi/pull/434/files#r48913804>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/434/files#r48913949>.
>